core-backend/src/funcs.rs

```diff
@@ -24,67 +24,309 @@ use crate::{
         TrapExplanation, UndefinedTerminationReason, UnrecoverableExecutionError,
         UnrecoverableMemoryError,
     },
-    memory::{MemoryAccessError, WasmMemoryRead},
+    memory::{ExecutorMemory, MemoryAccessError, WasmMemoryRead},
     runtime::CallerWrap,
+    state::HostState,
     BackendExternalities,
 };
 use alloc::string::{String, ToString};
 use blake2_rfc::blake2b::blake2b;
-use core::marker::PhantomData;
+use core::{fmt, marker::PhantomData};
 use gear_core::{
     buffer::{RuntimeBuffer, RuntimeBufferSizeError},
     costs::RuntimeCosts,
     env::{DropPayloadLockBound, Externalities},
     gas::CounterType,
+    ids::{MessageId, ProgramId},
     message::{
         HandlePacket, InitPacket, MessageWaitedType, Payload, PayloadSizeError, ReplyPacket,
     },
     pages::{PageNumber, PageU32Size, WasmPage},
 };
-use gear_core_backend_codegen::host;
 use gear_core_errors::{MessageError, ReplyCode, SignalCode};
-use gear_sandbox_env::HostError;
+use gear_sandbox::{ReturnValue, Value};
+use gear_sandbox_env::{HostError, WasmReturnValue};
 use gsys::{
     BlockNumberWithHash, ErrorBytes, ErrorWithBlockNumberAndValue, ErrorWithGas, ErrorWithHandle,
     ErrorWithHash, ErrorWithReplyCode, ErrorWithSignalCode, ErrorWithTwoHashes, Hash,
     HashWithValue, TwoHashesWithValue,
 };
 
-#[macro_export(local_inner_macros)]
-macro_rules! syscall_args_trace {
-    ($val:expr) => {
-        {
-            let s = ::core::stringify!($val);
-            if s.ends_with("_ptr") {
-                alloc::format!(", {} = {:#x?}", s, $val)
-            } else {
-                alloc::format!(", {} = {:?}", s, $val)
-            }
+const PTR_SPECIAL: u32 = u32::MAX;
+
+struct ValueFormatter<'a>(&'a Value);
```

Comment:
these are awesome, but please separate into special module, because its better to keep `mod funcs` as small and readable/easy-to-edit as possible